game/src/server/rooms/Room.ts

```diff
@@ -1,41 +1,75 @@
-import { Socket } from "socket.io";
+import { Namespace, Socket } from "socket.io";
 import { HostChangeEvent, UsersInfoSentEvent } from "../../communication/race/DataInterfaces";
 import { WAITING_ROOM_EVENT_NAMES } from "../../communication/race/EventNames";
 import UserInfo from "../../communication/userInfo";
-import GameFSM from "../../gameCore/gameState/GameFSM";
-import { GameState } from "../../gameCore/gameState/State";
+import { ServerGame } from "../../gameCore/Game";
+import State, { GameState } from "../../gameCore/gameState/State";
+import GameRepository from "../data/GameRepository";
 import User from "../data/User";
 
-export class Room {
+/**
+ * This class is a final state machine that represents the current state of the room. It is basically the container that will hold each game
+ * states like PreGame and RaceGame. Once a state ends, it will transfer to the next state and continue its operations like usual depending
+ * on the state it is currently using.
+ */
+export default class Room {
 	private max_player_count = 6;
-	private id: string;
-	private nsp: SocketIO.Namespace;
+	private readonly id: string;
+	private state: State;
 	private readonly roomString: string;
 	private users: User[] = [];
-	private gameFSM: GameFSM;
+	private nsp: SocketIO.Namespace;
+	private gameRepo: GameRepository;
 	private host: User;
 
-	constructor(id: string, nsp: SocketIO.Namespace, gameFSM: GameFSM, roomString: string) {
+	constructor(id: string, state: State, gameRepo: GameRepository, roomString: string, nsp: Namespace) {
 		this.id = id;
-		this.nsp = nsp;
-		this.gameFSM = gameFSM;
 		this.roomString = roomString;
+		this.nsp = nsp;
+		this.gameRepo = gameRepo;
+		//setting up starting state
+		this.transitionTo(state);
 	}
 
-	public getRoomId(): string {
+	public getId(): string {
 		return this.id;
 	}
 
+	public getUsers(): User[] {
+		return this.users;
+	}
+
+	public getRoomString(): string {
+		return this.roomString;
+	}
+
+	public getNamespace(): SocketIO.Namespace {
+		return this.nsp;
+	}
+
 	public getGameState(): GameState {
-		return this.gameFSM.getGameState();
+		return this.state.getStateType();
+	}
+
+	public transitionTo(nextState: State): void {
+		this.state = nextState;
+		this.state.setContext(this);
+	}
+
+	public gameStarted(game: ServerGame): void {
```

Comment:
Je changerais ça pour "startGame".

Dans pregame, on fait "**startRaceGame**".
Dans "startRaceGame", on fait "this.context.**gameStarted**"
Dans "gameStarted", on fait "this.gameRepo.**addGame**(game)".

J'pense que ça serait plus parlant si ça faisait : "pregame.startRaceGame" --> "Room.startGame" --> "GameRepo.addGame"

Reply:
Ça me fait penser, dans Pregame.startRaceGame, on fait : 
this.context.gameStarted(raceGame);
this.context.transitionTo(raceGame);

Pourquoi est-ce qu'on ferait pas juste : this.context.startGame(raceGame); et dans "startGame" on aurait la transition? J'vois pas de situation où on starterait une game sans faire de transition de state.